_posts/2015-09-01-Cocoa-Architecture-Dropped-Design-Patterns.markdown

```diff
@@ -0,0 +1,148 @@
+
+The part that eventually became a strong enough code-smell to warrant a re-write was the view controller which allowed a partner to choose what information to send would pass details individually to an object that would generate the HTML. I found it difficult to write simple tests for the class' behavior. Initially I would mock the email composer, then inspect the methods that were called. This felt wrong, as you shouldn't really be mocking classes you own. Given the importance of the functionality that classes provide our application, ideas on ways to improve the section of code stayed on my mind for a long time.
+
+The fix came to me during Justin Searl's talk [Sometimes a Controller is Just a Controller](https://speakerdeck.com/searls/sometimes-a-controller-is-just-a-controller) - specifically slide [55](https://speakerdeck.com/searls/sometimes-a-controller-is-just-a-controller?slide=55). Jon talks about objects that _either hold and describe a value or perform useful behavior, never both_.
+
+I took this advice and re-evaluated the relationship between settings view controller and composer object. Initially settings would configured the composer, afterwards the settings would create a configuration object and the composer would consume it. This made it _significantly_ easier to write tests for both objects. As they had very obvious inputs and outputs in the form of a [AREmailSettings](https://github.com/artsy/energy/blob/aa97d90cf37932d4c0f49ea4c4d31f7e491f16a6/Classes/Util/Emails/AREmailSettings.h). The [AREmailComposerTests](https://github.com/artsy/energy/blob/aa97d90cf37932d4c0f49ea4c4d31f7e491f16a6/ArtsyFolio%20Tests/Util/AREmailComposerTests.m) in particular become much more elegant.
+
+### Direct use of Responder Chain
+
+Before I worked at Artsy, I was a [Mac developer](http://i.imgur.com/Am9LjED.gif), I've been doing that since before iOS existed, so this influences my code style. One of the great parts of the Cocoa toolchain is the [responder chain](https://developer.apple.com/library/ios/documentation/EventHandling/Conceptual/EventHandlingiPhoneOS/event_delivery_responder_chain/event_delivery_responder_chain.html), a well documented way of passing methods up a known chain of objects. It solves a common problem with complicated view structures. You could have a button that is generated at runtime deep inside a view hierarchy and you would like the view controller to handle it being tapped. You could use a long chain of delegate methods, or use a [private method](https://twitter.com/unimp0rtanttech/status/555828778015129600) to get the reference to the view controller instance. On the Mac usage of the responder chain is a common pattern, on iOS it is used rarely.
```

Comment:
> You could have a button that is generated at runtime deep inside a view hierarchy and you would like the view controller to handle it being tapped.

Does NSViewController actually properly join in the chain nowadays? Because it definitely used to not join in at all.